### PR TITLE
Implement and test new version of process fidelity

### PR DESCRIPTION
**Pull request by fhopfmueller**
A re-implementation of `qutip.metrics.process_fidelity`. 
The definition implemented here follows Gilchrist, Langford, Nielsen, Phys. Rev. A 71, 062310 (2005), https://journals.aps.org/pra/abstract/10.1103/PhysRevA.71.062310 (https://arxiv.org/abs/quant-ph/0408063).
Depending on how the quantum channels passed to `process_fidelity` are represented (unitary, supermatrix, choi, chi, or Kraus), computational shortcuts are used. They are all tested against each other.

**Related issues or PRs**
Proposed in https://github.com/qutip/qutip/issues/1703

**Changelog**
Fixed implementation of qutip.metrics.process_fidelity and added tests

### CODE REVIEW

**Review by Ericgig — COMMENTED** [review]
Hi fhopfmueller,
Thank you for the contribution.

Adding support for different representation of super, tests and a proper documentation with a reference is great.

But changing the definition and interface of an existing function could cause issue to those using it. This kind of change would be more appropriate for `dev.major`.

The code is quite hard to follow with a lot of branches and returns. You could use a private function for the case where there are no `target` to split the function into 2 and make it easier to read.

**Comment by fhopfmueller** [comment]
Hi Eric,

Thanks for looking at this and your feedback!

Alright, I'll close this PR and submit a new one for `dev.major`, and will split up the function as you suggest.
